### PR TITLE
fix(server): handle decodeURI errors

### LINKED ISSUE #4977: URI malformed
### Version

[v2.4.3](https://github.com/nuxt.js/releases/tag/v2.4.3)

### Reproduction link

[https://nuxtjs.org/%c1%81](https://nuxtjs.org/%c1%81)

### Steps to reproduce

1) Load https://nuxtjs.org/%c1%81
2) See console error

### What is expected ?

Trigger 404/500 errors

### What is actually happening?

Execution stops

### Additional comments?

(node:87717) UnhandledPromiseRejectionWarning: URIError: URI malformed
    at decodeURI (<anonymous>)
    at nuxtMiddleware (node_modules/@nuxt/server/dist/server.js:120:15)

Perhaps the decodeURI should be wrapped in the try {} block?
https://github.com/nuxt/nuxt.js/blob/dev/packages/server/src/middleware/nuxt.js#L10

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8619">#c8619</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by phof**
## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Resolves #4977

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by phof** [comment]
@manniL I've included a try/catch for the `decodeURI` function.

I'm not immediately clear if there's a way to pass the error's name/message to the server-side error template, as this should result in a "Client Error" (400 response status), not a "Server Error".

File: packages/server/src/middleware/nuxt.js [hunk 1]
```diff
@@ -7,7 +7,22 @@ import { getContext } from '@nuxt/utils'
...
   // Get context
   const context = getContext(req, res)
-  const url = decodeURI(req.url)
+  let url
+
+  try {
+    url = decodeURI(req.url)
+  } catch (err) {
+    /* istanbul ignore if */
+    if (context && context.redirected) {
```
Comment (manniL) [inline packages/server/src/middleware/nuxt.js]: You probably don't need that block here as it'll be caught further down the road.

File: packages/server/src/middleware/nuxt.js [hunk 2]
```diff
@@ -7,7 +7,22 @@ import { getContext } from '@nuxt/utils'
...
+      consola.error(err)
+      return err
+    }
+
+    const e = new Error('URL malformed')
```
Comment (manniL) [inline packages/server/src/middleware/nuxt.js]: I'd love to see a more descriptive variable name here :relaxed:

Comment (manniL) [reply packages/server/src/middleware/nuxt.js]: Maybe you could use the error coming from the failed decoding directly anyway? :thinking: 

(see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURI#Catching_errors)

**Review by TheAlexLichter — CHANGES_REQUESTED** [review]

File: packages/server/src/middleware/nuxt.js (hunk 2 above)
Comment (phof) [reply packages/server/src/middleware/nuxt.js]: Moved everything in the same `try` block and setting `statusCode = 400` in the `catch`.

**Review by pi0 — APPROVED** [review]

**Comment by japboy** [comment]
this change prevents to control application level error handling for malformed urls and forces me to show error pages to users...

**Pull request merged**